src/deps/httparse/mod.rs

```diff
@@ -0,0 +1,1282 @@
+    ///
+    /// A header name must be valid ASCII-US, so it's safe to store as a `&str`.
+    pub name: &'a str,
+    /// The value portion of a header.
+    ///
+    /// While headers **should** be ASCII-US, the specification allows for
+    /// values that may not be, and so the value is stored as bytes.
+    pub value: &'a [u8],
+}
+
+/// An empty header, useful for constructing a `Header` array to pass in for
+/// parsing.
+pub const EMPTY_HEADER: Header<'static> = Header { name: "", value: b"" };
+
+#[inline]
+fn parse_version(bytes: &mut Bytes) -> Result<u8> {
+    if let Some(mut eight) = bytes.next_8() {
+        expect!(eight._0() => b'H' |? Err(Error::Version));
+        expect!(eight._1() => b'T' |? Err(Error::Version));
+        expect!(eight._2() => b'T' |? Err(Error::Version));
+        expect!(eight._3() => b'P' |? Err(Error::Version));
+        expect!(eight._4() => b'/' |? Err(Error::Version));
+        expect!(eight._5() => b'1' |? Err(Error::Version));
+        expect!(eight._6() => b'.' |? Err(Error::Version));
+        let v = match eight._7() {
+            b'0' => 0,
+            b'1' => 1,
+            _ => return Err(Error::Version)
+        };
+        return Ok(Status::Complete(v))
+    }
+
+    // else (but not in `else` because of borrow checker)
+
+    // If there aren't at least 8 bytes, we still want to detect early
+    // if this is a valid version or not. If it is, we'll return Partial.
+    expect!(bytes.next() == b'H' => Err(Error::Version));
+    expect!(bytes.next() == b'T' => Err(Error::Version));
+    expect!(bytes.next() == b'T' => Err(Error::Version));
+    expect!(bytes.next() == b'P' => Err(Error::Version));
+    expect!(bytes.next() == b'/' => Err(Error::Version));
+    expect!(bytes.next() == b'1' => Err(Error::Version));
+    expect!(bytes.next() == b'.' => Err(Error::Version));
+    Ok(Status::Partial)
+}
+
+/// From [RFC 7230](https://tools.ietf.org/html/rfc7230):
+///
+/// > ```notrust
+/// > reason-phrase  = *( HTAB / SP / VCHAR / obs-text )
+/// > HTAB           = %x09        ; horizontal tab
+/// > VCHAR          = %x21-7E     ; visible (printing) characters
+/// > obs-text       = %x80-FF
+/// > ```
+///
+/// > A.2.  Changes from RFC 2616
+/// >
+/// > Non-US-ASCII content in header fields and the reason phrase
+/// > has been obsoleted and made opaque (the TEXT rule was removed).
+///
+/// Note that the following implementation deliberately rejects the obsoleted (non-US-ASCII) text range.
+///
+/// The fully compliant parser should probably just return the reason-phrase as an opaque &[u8] data
+/// and leave interpretation to user or specialized helpers (akin to .display() in std::path::Path)
+#[inline]
+fn parse_reason<'a>(bytes: &mut Bytes<'a>) -> Result<&'a str> {
+    loop {
+        let b = next!(bytes);
+        if b == b'\r' {
+            expect!(bytes.next() == b'\n' => Err(Error::Status));
+            let res = str::from_utf8(bytes.slice_skip(2)).map_err(|_e| Error::Status)?;
```

Comment:
Are these utf8 parse errors possible? I think this loop ensures that all the bytes are ASCII, so these could just be unwraps instead of `map_err`s.

Reply:
Some background: the `httparse` directory is a copy of https://github.com/seanmonstar/httparse, but with the unsafe parts removed.

Generally speaking, I'm leery of `unwrap`s in network-facing code.  Malicious network requests should never cause panics, so I'd prefer `map_err` if at all possible.

Reply:
Yep, in the original code, these are unchecked casts to a utf8 string. If an error occurs here, there is something wrong with the http parsing code.

Reply:
I'd still rather leave this as a recoverable error that leads to the connection being aborted instead of leading to the node itself crashing.

Reply:
Works for me